flink/src/main/java/org/apache/iceberg/flink/sink/RowDataTaskWriter.java

```diff
@@ -0,0 +1,141 @@
+  private final PartitionKey partitionKey;
+  private final RowDataWrapper rowDataWrapper;
+  private final List<Integer> equalityFieldIds;
+
+  private final Map<PartitionKey, RowDataDeltaWriter> deltaWriterMap = Maps.newHashMap();
+
+  RowDataTaskWriter(PartitionSpec spec,
+                    FileFormat format,
+                    FileAppenderFactory<RowData> appenderFactory,
+                    OutputFileFactory fileFactory,
+                    FileIO io,
+                    long targetFileSize,
+                    Schema schema,
+                    RowType flinkSchema,
+                    List<Integer> equalityFieldIds) {
+    super(spec, format, appenderFactory, fileFactory, io, targetFileSize);
+
+    this.schema = schema;
+    this.equalityFieldIds = equalityFieldIds;
+
+    this.partitionKey = new PartitionKey(spec, schema);
+    this.rowDataWrapper = new RowDataWrapper(flinkSchema, schema.asStruct());
+  }
+
+  @Override
+  public void write(RowData row) throws IOException {
+    RowDataDeltaWriter deltaWriter;
+
+    if (spec().fields().size() <= 0) {
+      // Create and cache the delta writer if absent for unpartitioned table.
+      deltaWriter = deltaWriterMap.get(UNPARTITIONED_KEY);
+      if (deltaWriter == null) {
+        deltaWriter = new RowDataDeltaWriter(null, equalityFieldIds);
+        deltaWriterMap.put(UNPARTITIONED_KEY, deltaWriter);
+      }
+    } else {
+      // Refresh the current partition key.
+      partitionKey.partition(rowDataWrapper.wrap(row));
+
+      // Create and cache the delta writer if absent for partitioned table.
+      deltaWriter = deltaWriterMap.get(partitionKey);
+      if (deltaWriter == null) {
+        // NOTICE: we need to copy a new partition key here, in case of messing up the keys in writers.
+        PartitionKey copiedKey = partitionKey.copy();
+        deltaWriter = new RowDataDeltaWriter(copiedKey, equalityFieldIds);
+        deltaWriterMap.put(copiedKey, deltaWriter);
+      }
+    }
+
+    switch (row.getRowKind()) {
+      case INSERT:
+      case UPDATE_AFTER:
+        deltaWriter.write(row);
+        break;
+
+      case DELETE:
+      case UPDATE_BEFORE:
+        deltaWriter.delete(row);
+        break;
+
+      default:
+        throw new UnsupportedOperationException("Unknown row kind: " + row.getRowKind());
+    }
+  }
+
+  @Override
+  public void close() {
+    Tasks.foreach(deltaWriterMap.values())
+        .throwFailureWhenFinished()
+        .noRetry()
+        .run(deltaWriter -> {
+          try {
+            deltaWriter.close();
+          } catch (IOException e) {
+            throw new UncheckedIOException(e);
+          }
+        });
+  }
+
+  private class RowDataDeltaWriter extends BaseDeltaWriter {
+
+    private RowDataDeltaWriter(PartitionKey partition, List<Integer> equalityFieldIds) {
+      super(partition, equalityFieldIds, schema);
+    }
+
+    @Override
+    protected StructLike asKey(RowData row) {
+      return rowDataWrapper.wrap(row);
```

Comment:
I think this is a bug. It doesn't extract the key fields, it just wraps the row as a `StructLike`. It should extract the equality fields to produce a key, probably using `StructProjection`.

Reply:
I think the tests work because the ID column is the first column in the record.

Reply:
The tests work because the `StructLikeMap`  has a `StructLikeWrapper` which will only compare the equality fields, even here we provided the full columns.  The name `asKey` and `asCopiedKey` are not inappropriate here,  `asStructLike` and `asCopiedStructLike` will be better.

Reply:
> The tests work because the StructLikeMap has a StructLikeWrapper which will only compare the equality fields

But this happens using the key schema and fields are accessed by position. Wouldn't that fail if the key schema wasn't a prefix of the row schema?